File: backend/packages/users/src/utils/router.py
import json
import traceback
import typing as t
from .logger import BaseLogger
from .route import BaseRoute
from .request import Request
from .response import Response


class Router:
    _logger: BaseLogger
    _routes: t.Dict[str, BaseRoute]

    def __init__(self, logger: BaseLogger) -> None:
        self._logger = logger
        self._routes = {}

    def register(self, routeKey: str, handler: BaseRoute):
        self._routes[routeKey] = handler

    def call(self, event):
        request = Request(event=event)
        response = None

        self._logger.info(
            f"[Connection id {request.connection_id}{'' if request.request_id is None else f', request id {request.request_id} -> request'}] Route key: {request.route}, data: {json.dumps(request.data)}"
        )

        if request.route not in self._routes:
            response = Response(
                status_code=404,
                action=request.action,
                request_id=request.request_id,
                data={},
            )
        else:
            try:
                response = self._routes[request.route].call(request)

            except Exception as e:
                self._logger.error(
                    f"[Connection id {request.connection_id}{'' if request.request_id is None else f', request id {request.request_id} !! error'}] Route key: {request.route}, error: {e}, traceback: {traceback.format_exc()}"
                )

                response = Response(
                    status_code=500,
                    action=request.action,
                    request_id=request.request_id,
                    data={},
                )

        if not response:
            return

        self._logger.info(
            f"[Connection id {request.connection_id}{'' if request.request_id is None else f', request id {request.request_id} <- response'}] Route key: {request.route}, status code: {response.status_code}, data: {json.dumps(response.data)}"
        )

        return response

File: backend/packages/data/src/utils/router.ts
import { Logger } from "./logger"
import { makeRequest } from "./request"
import { Response, makeResponse } from "./response"
import { BaseRoute } from "./route"

export interface Router {
    register<Input, Output>(routeKey: string, handler: BaseRoute<Input, Output>): void
    call<Output>(event: any): Promise<Response<Output> | Response<undefined>>
}

type Self = {
    logger: Logger
    routes: Record<string, BaseRoute<any, any>>
}

export const makeRouter = (logger: Logger): Router => {
    const self: Self = {
        logger,
        routes: {}
    }

    return {
        register: register(self),
        call: call(self)
    }
}

const register = (self: Self): Router['register'] => (routeKey, handler) => {
    self.routes[routeKey] = handler
}

const call = (self: Self): Router['call'] => async (event) => {
    const request = makeRequest(event)
    let response: Response<any>

    self.logger.info(
        `[Connection id ${request.connectionId}${request.requestId === undefined ? '' : `, request id ${request.requestId} -> request`}] Route key: ${request.route}, data: ${JSON.stringify(request.data)}`
    )

    const routeHandler = self.routes[request.route]

    if (!routeHandler) {
        response = makeResponse(404, request.action, request.requestId, { "reason": "Not found" })
    } else {
        try {
            response = await routeHandler.call(request)
        } catch (error) {
            self.logger.error(
                `[Connection id ${request.connectionId}${request.requestId === undefined ? '' : `, request id ${request.requestId} !! error`}] Route key: ${request.route}, error: ${(error as Error).message}, traceback: ${(error as Error).stack}`
            )

            response = makeResponse(
                500,
                request.action,
                request.requestId,
                { reason: "Internal server error" }
            )
        }
    }

    if (response === undefined) {
        response = makeResponse(
            200,
            undefined,
            undefined,
            undefined
        )
    }

    self.logger.info(
        `[Connection id ${request.connectionId}${request.requestId === undefined ? '' : `, request id ${request.requestId} <- response`}] Route key: ${request.route}, status code: ${response.statusCode}, ${!('data' in response) ? '' : `data: ${JSON.stringify(response.data)}`}`
    )

    return response
}
